lib/Optimizer/Dialect/CC/CCOps.cpp

```diff
@@ -200,9 +200,46 @@ void cudaq::cc::AllocaOp::getCanonicalizationPatterns(
 //===----------------------------------------------------------------------===//
 
 OpFoldResult cudaq::cc::CastOp::fold(FoldAdaptor adaptor) {
-  // If cast is a nop, just forward the argument to the uses.
-  if (getType() == getValue().getType())
-    return getValue();
+  // Replace unnecessary casts.
+  if (getType() == getValue().getType()) {
+    auto val = getValue();
+    auto ty = getType();
+
+    // Cast is a nop, just forward the argument to the uses.
+    if (!isa<IntegerType>(ty))
+      return val;
+
+    // Cast that does not change the type or the sign.
+    if (!getZint() && !getSint())
+      return val;
+
+    // Cast of a boolean to a boolean.
+    if (ty.getIntOrFloatBitWidth() == 1)
+      return val;
+
+    // Cast of a known unsigned value to an unsigned value.
+    if (val.getDefiningOp<cc::StdvecSizeOp>() ||
+        val.getDefiningOp<quake::VeqSizeOp>())
+      if (getZint())
+        return val;
+
+    // Fold consecutive casts when possible.
```

Comment:
Eliminating cast chains is already done someplace ...

Reply:
This one belongs in FuseCastCascade.

Reply:
will do, thanks!

Reply:
done!